### DISCUSSION TITLE
bug in FastTMXTiledMap if visibility off in Tiled?

### DISCUSSION THREAD

**Discussion by rudiHammad** [axmolengine/axmol#1281] [Q&A] [answered]
Hi, 
I noticed something that seems like a bug. I made a video about it. If I save the tmx file with the layer visibility ON in Tiled, everything works fine. But my layer visibility is OFF and I save the tmx file axmol crashed saying that the layer is null as the error shows `this->**_metaLayer**` was null. This is strange because I checked the tmx file and the data is there. 
The data with the visibility of the layer "meta" on is  ` <layer id="3" name="meta" width="31" height="17"> `  this works.

The data with the visibility of the layer "meta" on is ` <layer id="3" name="meta" width="31" height="17" visible="0"> ` , this crashes.

https://github.com/axmolengine/axmol/assets/25636976/459ea557-2eba-4df4-bff0-c69874c8e774

**Comment by aismann** [comment]
@rudiHammad 
Have you an example source which shows this "behavior"?
A video clip is mostly a nice look but is hard to repeat the issue for debuging on source

**Reply by rudiHammad** [reply]
I'll try to upload something when I have some time, but it is pretty straightforward. If you are familiar with tiled, you create a simple map, with 2 layers: the one with the render sprites and the one with the collidable tiles called meta (these tiles have a bool property called "collidable").
Then the only part of the code you need is this:
```
bool HelloWorld::init()
{
    //////////////////////////////
    // 1. super init first
    if (!Scene::init())
    {
        return false;
    }

    auto visibleSize = _director->getVisibleSize();
    auto origin = _director->getVisibleOrigin();
    auto safeArea = _director->getSafeAreaRect();
    auto safeOrigin = safeArea.origin;

 // Creating the TileMap and accessing its data (tmx)
    _tileMap = new ax::FastTMXTiledMap();    // _tileMap being a private header member
    _tileMap->initWithTMXFile("tiled/tilesmap.tmx");
    this->addChild(_tileMap);

  _metaLayer = _tileMap->getLayer("meta");      // _metaLayer being a private header member
 _metaLayer->setVisible(false);  // if the tmx file has the meta layer visibility off, this line crashes because it thinks that layer doesn't exists, // so _metaLayer is null 

}

```

I'll try to upload when I have sometime a the tilemap project

**Reply by rh101** [reply, marked as answer]
Are you sure this is a bug? According to the current implementation, if the layer visibility is `false`, then it doesn't get created.

```
void FastTMXTiledMap::buildWithMapInfo(TMXMapInfo* mapInfo)
{
...
    auto& layers = mapInfo->getLayers();
    for (const auto& layerInfo : layers)
    {
        if (layerInfo->_visible) <<<< HERE >>>>
        {
            FastTMXLayer* child = parseLayer(layerInfo, mapInfo);
            if (child == nullptr)
            {
                idx++;
                continue;
            }
            addChild(child, idx, idx);
...
        }
    }
...
}
```

So, when you select a layer via `getLayer()`, you should always check the return value. If it's null, then it doesn't exist in the current map.  That may have been implemented like that on purpose.

**Reply by rudiHammad** [reply]
hmm..I see.
I think that could be a workflow issue. If you are an artist working in Tiled, you might have many layers of data and only one to render.
So when ever you save the tmx file you are forced to set all those layer visible. Then if you want to continue workin in Tiled, you have to hide again all the ones you don't need. Imagine you have 20 layers and you specifically working on 1 that has collectable tiles. When ever you are saving the tmx, you have to unhide the other 19, and then in axmol get all those 19 layer in a loop and set their visibility to false.
Then go back to Tile, hide again the 19 layers you are not working on, etc...
It would be much easier to automatically create the layers that are not visible and set then their visibility to false.

I don't know, maybe there is a reason as you said to be set that way, but I am not sure why.

**Reply by rh101** [reply]
> I think that could be a workflow issue. If you are an artist working in Tiled, you might have many layers of data and only one to render.
> So when ever you save the tmx file you are forced to set all those layer visible.

I understand.

> I don't know, maybe there is a reason as you said to be set that way, but I am not sure why.

Yeah, I'm not entirely sure why it was implemented in this way; perhaps for performance reasons, but that's only a guess.  You can always modify the code to ensure the layer gets created, and then set the node visibility to false, and see what happens (in terms of performance or other issues that may appear).

**Comment by rudiHammad** [comment]
It's okey, I prefer not to change axmol. I might do an exportation tool in Tiled.
Cheers,
R
